### ISSUE TITLE
[UX] Change the password reset form to use a single password field + show/hide toggle

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#4440]
This was brought up in https://github.com/backdrop/backdrop-issues/issues/4281#issuecomment-578466760

@stpaultim 
>I notice that we only ask for a new password once when a user creates a new account. We only ask for a a new password once when an authenticated user chooses to change their password.
>
>Should we ask for it twice when a user uses password reset to create a new password.
>
>I don't feel strongly about this. But, it seems like Backdrop CMS has already made a decision that we don't want to require the double password thing. So, why not be consistent.

@indigoxela 

> Good point. As we're already working on it, why not switch to a single field?
> ```php
>   $form['pass'] = array(
>     '#type' => 'password',
>     '#password_toggle' => TRUE,
>     '#password_strength' => TRUE,
>     '#required' => TRUE,
>   );
> ```
> ...there's a concern: Changing the form field type could be a contrib-breaker, for instance for passphrase_policy module. This has to be evaluated, though.
> 
> Setting the same field to required is a task for a simple bugfix release, changing the form item type requires more evaluation and probably isn't appropriate for a bugfix release.
> 
> @stpaultim I'd rather recommend to create a new issue for the form field type change.

**Comment by klonos** [comment]
> I'd rather recommend to create a new issue for the form field type change.

Here it is 😉

**Comment by klonos** [comment]
PR up for review. I have tested passphrase_policy with lates 1.x on my local, and with 1.1.3 in the PR sandbox, and have found no obvious issues. I also reviewed `passphrase_policy_form_alter()` (https://github.com/backdrop-contrib/passphrase_policy/blob/1.x-1.x/passphrase_policy.module#L81) and I see nothing obvious that could be affected by the change.

More testing welcome, but I think that we shouldn't worry too much, given:

- the very low, and declining usage of the module: https://backdropcms.org/project/usage/passphrase_policy
- the fact that we plan to ship Backdrop with some basic policy restrictions (#4265) makes me believe that there will be less need for something custom/contrib.

**Comment by indigoxela** [comment]
@klonos sorry, currently there are conflicts in your PR that need to get resolved. Mind to update your code?

**Comment by herbdool** [comment]
This is too big a change for a bug release.

**Comment by klonos** [comment]
Sorry for taking so long to respond here 😞 

Rebased, and tests are green. Ready for testing/review again.

**Comment by bugfolder** [comment]
Started to test and review, had to close and reopen the PR to get the Tugboat site to rebuild, and now tests aren't passing (same test fail on all 3 PHPs, so it's not the gremlin this time). 😞

**Comment by yorkshire-pudding** [comment]
I closed and re-opened the PR to get new tests.  Here are the failed test results:

<details>
<summary>Test failures</summary>

```
Detailed test results
---------------------
---- User: User login (UserLoginTestCase) ----

Status    Group      Filename          Line    Function                            
------------------------------------------------------------------------------------------------------------------------
Fail      Other      user.test         671     resetUserPassword()                                                     
    Failed to set field pass[pass1] to K8NPHmZ4ag
Fail      Other      user.test         671     resetUserPassword()                                                     
    Failed to set field pass[pass2] to K8NPHmZ4ag
Fail      Other      user.test         671     resetUserPassword()                                                     
    Found the requested form fields at
Fail      Other      user.test         671     resetUserPassword()                                                     
    Failed to set field pass[pass1] to UoxrDtuMhc
Fail      Other      user.test         671     resetUserPassword()                                                     
    Failed to set field pass[pass2] to UoxrDtuMhc
Fail      Other      user.test         671     resetUserPassword()                                                     
    Found the requested form fields at
Fail      User login user.test         562     testPerUserLoginFloodControl()                                          
    User wvdy94oe successfully logged in.
Fail      Other      user.test         563     testPerUserLoginFloodControl()                                          
    Raw "Member for" found

---- User: Reset password (UserPasswordResetTest) ----

Status    Group      Filename          Line    Function                            
------------------------------------------------------------------------------------------------------------------------
Fail      Other      user_password_res 269     testPasswordResetFloodControlPerUser()                                  
    Failed to set field pass[pass1] to n6EEbPOi0BWfUCjem0Kp
Fail      Other      user_password_res 269     testPasswordResetFloodControlPerUser()                                  
    Failed to set field pass[pass2] to n6EEbPOi0BWfUCjem0Kp
Fail      Other      user_password_res 269     testPasswordResetFloodControlPerUser()                                  
    Found the requested form fields at
Fail      Other      user_password_res 276     testPasswordResetFloodControlPerUser()                                  
    Password reset instructions mailed message displayed.
Fail      Other      user_password_res 278     testPasswordResetFloodControlPerUser()                                  
    Flood control was not triggered by password reset.
Fail      Other      user_password_res 276     testPasswordResetFloodControlPerUser()                                  
    Password reset instructions mailed message displayed.
Fail      Other      user_password_res 278     testPasswordResetFloodControlPerUser()                                  
    Flood control was not triggered by password reset.
Fail      Other      user_password_res 405     testUserPasswordResetAlreadyLoggedIn()                                  
    Failed to set field pass[pass1] to bWGVKpU7cR
Fail      Other      user_password_res 405     testUserPasswordResetAlreadyLoggedIn()                                  
    Failed to set field pass[pass2] to bWGVKpU7cR
Fail      Other      user_password_res 405     testUserPasswordResetAlreadyLoggedIn()                                  
    Found the requested form fields at
Fail      Other      user_password_res 407     testUserPasswordResetAlreadyLoggedIn()                                  
    "Your account password has been updated." found
Error: Process completed with exit code 1.
```

</details>

TLDR:

These test cases need updating.

- User: User login (UserLoginTestCase)
  - resetUserPassword()
  - testPerUserLoginFloodControl()
- User: Reset password (UserPasswordResetTest)
  - testPasswordResetFloodControlPerUser()
  - testUserPasswordResetAlreadyLoggedIn(

**Comment by yorkshire-pudding** [comment]
By following [the excellent instructions on how to fork a PR](https://backdropcms.org/news/how-to-fork-a-pr-in-the-issue-queue) provided by @stpaultim I have taken the PR by @klonos and made some minor updates to the tests so they work with the new form structure and provided a new PR.

I am happy to ADVOCATE for this change.

**Comment by herbdool** [comment]
I've tested and reviewed. It's working well. I too don't see any conflicts with other modules, including passphrase_policy.

**Comment by quicksketch** [comment]
Thank you @yorkshire-pudding, @klonos, @herbdool, @indigoxela, and @bugfolder! I merged https://github.com/backdrop/backdrop/pull/5094 into 1.x for 1.32.0.
